File: solutions/day8.R
execute <- function(grid, instructions) {
    parse <- function(line) {
        parts <- strsplit(line, "\\s") |>
            unlist()
        if (parts[[1]] == "rect") {
            bounds <- strsplit(parts[[2]], "x") |>
                unlist() |>
                strtoi()
            function() rect(bounds[[1]], bounds[[2]])
        } else {
            target <- strsplit(parts[[3]], "=") |>
                unlist() |>
                tail(1) |>
                strtoi()
            n <- strtoi(parts[[5]])
            func <- if (parts[[2]] == "row") rotate_row else rotate_column
            function() func(target + 1, n)
        }
    }


    rotate_vector <- function(x, n) {
        len <- length(x)
        cutpoint <- len - n + 1
        right <- x[cutpoint:len]
        left <- x[1:(cutpoint - 1)]
        c(right, left)
    }

    rect <- function(x, y) {
        grid[1:y, 1:x] <<- 1
    }


    rotate_row <- function(row, n) {
        grid[row, ] <<- rotate_vector(grid[row, ], n)
    }

    rotate_column <- function(col, n) {
        grid[, col] <<- rotate_vector(grid[, col], n)
    }

    instructions <- lapply(raw_input, parse)
    for (line in instructions) {
        line()
    }
    grid
}
raw_input <- readLines("inputs/day8.txt")
nrow <- 6
ncol <- 50
grid <- matrix(0, nrow = nrow, ncol = ncol)
grid <- execute(grid, raw_input)

part1 <- sum(grid)
print(part1)

apply(grid, MARGIN = 2, FUN = rev) |>
    t() |>
    image()

File: solutions/day8.py
import re

import matplotlib.pyplot as plt
import numpy as np
from utils.utils import split_lines

raw_input = split_lines("inputs/day8.txt")

grid = np.zeros([6, 50])
word_pattern = re.compile(r"[a-z]{3,}")


def rect(x, y):
    grid[:y, :x] = np.ones([y, x])


def rotate_row(row, n):
    grid[row, :] = np.roll(
        grid[row, :],
        n,
    )


def rotate_column(col, n):
    grid[:, col] = np.roll(grid[:, col], n)


mapping = {"rotate_row": rotate_row, "rotate_column": rotate_column, "rect": rect}


def process_line(line):
    func = "_".join(re.findall(word_pattern, line))
    args = [int(x) for x in re.findall(r"\d+", line)]
    return {"func": func, "args": args}


processed = [process_line(line) for line in raw_input]

for line in processed:
    mapping[line["func"]](*line["args"])

part1 = np.sum(grid)
print(part1)

plt.imshow(grid)
plt.colorbar()
plt.show()
